### PR TITLE
Setting CASConcurrency to global accross parallel requests.

### PR DESCRIPTION
**Pull request by ola-rozenfeld**
Also limiting the default to 50 uploaders. This is the first step in the direction of limiting the overall memory (and network throughput) used by simultaneous uploads.

### CODE REVIEW

**Comment by ola-rozenfeld** [comment]
This way to do it is sub-optimal for a few reasons:
1. Retries are executed on the same uploader thread. This means, if 50 uploads are executing, 100 are waiting, and one failed, that failed one will use the thread to wait for an exponentially increasing period of time, while the 100 continue to wait. However, doing this thing optimally is not easy!
2. It doesn't remember which digests are concurrently being uploaded by other threads.
3. It doesn't remember which digests have already been uploaded.

But all these things have a lower priority than getting the memory limited in the first place.

File: go/client/client.go [hunk 1]
```diff
@@ -61,12 +61,12 @@ type Client struct {
...
-	Retrier        *Retrier
-	chunkMaxSize   ChunkMaxSize
-	useBatchOps    UseBatchOps
-	casConcurrency CASConcurrency
-	rpcTimeout     time.Duration
-	creds          credentials.PerRPCCredentials
+	Retrier      *Retrier
+	chunkMaxSize ChunkMaxSize
+	useBatchOps  UseBatchOps
+	casUploaders chan bool
```
Comment (ramymedhat) [inline go/client/client.go]: I think semaphores are a more natural fit here.

Comment (ola-rozenfeld) [reply go/client/client.go]: Not according to this: http://jmoiron.net/blog/limiting-concurrency-in-go/ which was the second result on controlling concurrency in Go. The first result (https://blog.golang.org/pipelines) agrees with this pattern as well. In general, the guideline is "you should rarely need to use sync package primitives, use channels instead".

File: go/client/client.go [hunk 2]
```diff
@@ -93,14 +93,13 @@ func (u UseBatchOps) Apply(c *Client) {
...
-// methods can result in N * CASConcurrency requests in flight at once.
+// CASConcurrency is the number of simultaneous requests that will be issued for CAS upload and
+// download operations.
 type CASConcurrency int
 
 // Apply sets the CASConcurrency flag on a client.
 func (cy CASConcurrency) Apply(c *Client) {
-	c.casConcurrency = cy
+	c.casUploaders = make(chan bool, cy)
 }
```
[the remark this answers is not in the document]
Comment (ola-rozenfeld) [reply go/client/client.go]: It's a design pattern. Look at the Opt interface which the CasConcurrency type implements, as well as all the other Client options. This way, we can pass variadic Opt to Dial (will be renamed to New at some point).

File: go/client/client.go (hunk 1 above)
Comment (ramymedhat) [reply go/client/client.go]: Discussed offline. sgtm.

**Pull request merged**